### PR TITLE
Monorepo

### LINKED ISSUE #60: [Feature] Switch to a monorepo
Several packages in a pkg directory (it conflicts less often with package.json when doing completions on the command line than packages)

- peggy (always/only published as a minified UMD which is tested for commonjs, es, ts, browser-script-src, and browser-import)
- peggy-cli (which depends on peggy and commander, and is node12+ starting in 2 weeks)
- sample programs (ts, etc), each marked `private` in their `package.json`

### PR DESCRIPTION
**Pull request by hildjj**
Fixes #60 

This looks like a massive set of changes, but it's mostly moving things around, switching to webpack, and making everything work.  I'm going to leave this as a draft until we all have a chance to look at it.

See https://github.com/peggyjs/peggy/projects/1 for follow-on work that comes after this.

### CODE REVIEW

**Comment by hildjj** [comment]
I still need to fixup the README's

**Comment by hildjj** [comment]
TODO: rewrite the `release` script

File: .github/workflows/node.js.yml
Comment (StoneCypher) [inline .github/workflows/node.js.yml:38]: in the future i think this should be a matrix build across versions of node and operating systems.

i can show you how to do this if you don't already know; it's easy.

i don't think this PR should block on this, so [I opened a new issue about it](https://github.com/peggyjs/peggy/issues/75) instead.

File: README.md
Comment (StoneCypher) [inline README.md:114]: i request that this say es2015, because there's an older forgotten es4x module format, and because it's likely that es9 or es10 will deliver an incompatible module format that exposes types and directly solves the michael jacksonscript problem

this also leaves it in parity with typescript, rollup, parcel, and rome

File: docs/development/test.html
Comment (StoneCypher) [inline docs/development/test.html:22]: This means that the testing repo has an external dependency off of Microsoft's network; this generates unnecessary traffic and slows your builds down ~5 seconds.  

If you install mocha as a dev dependency instead, you will get these from Microsoft's internal package re-host, which is much much faster.

It also means that as Microsoft begins to engage security on their package re-host, you aren't left out.

File: docs/development/test.html
Comment (StoneCypher) [inline docs/development/test.html:54]: I strongly advise that you consider `jest` rather than `mocha` as a test runner

`mocha` is largely dead and its coverage is quite seriously broken

I'm willing to do the conversion

I just converted a set of >2,000 `ava` tests to `jest` [a few months ago](https://github.com/StoneCypher/jssm/pull/417), without using codemods.  I understand how much work I'm getting myself into by offering this.

It's not worth blocking this PR, so, [I made another post-PR issue](https://github.com/peggyjs/peggy/issues/76).

File: docs/js/test-bundle.min.js.LICENSE.txt
Comment (StoneCypher) [inline docs/js/test-bundle.min.js.LICENSE.txt:581]: side note, unimportant: i don't know what this is, and if someone else does, i'm quite curious

File: package.json
Comment (StoneCypher) [inline package.json:2]: This name change will cause havoc in `npm`.  Many downstream tools believe that the repo name and the listing name must match.

File: package.json
Comment (StoneCypher) [inline package.json:5]: wut

File: package.json
Comment (StoneCypher) [inline package.json:22]: Please do not require downstream users to install a secondary package manager to rebuild.  This was a problem with futagoza: many people, myself included, legally cannot install this in many contexts, meaning our ability to patch then test in context is severely impeded.

This monorepo strategy is going to cause damage and does not appear to deliver actual value.

Can we please just do it the normal way 😥  It's so much simpler and doesn't require third party tools that exclude other developers

File: package.json
Comment (StoneCypher) [inline package.json:25]: as above.  this means i also can't test on work machines 😥

there is no actual value being delivered here.  this can be done without `pnpm`

there is no apparent need to subdivide the base repo.  

this will lead to constant conway churn

File: package.json
Comment (StoneCypher) [inline package.json:35]: this excludes me from participation without apparent actual value

File: pkg/benchmarks/.eslintrc.js
Comment (StoneCypher) [inline pkg/benchmarks/.eslintrc.js:5]: this isn't actually correct.  peg is expected to run in browsers and free engines, not just in node.

this disables eslint's ability to warn you when you're assuming node tools exist, which often won't.

File: pkg/benchmarks/.eslintrc.js
Comment (StoneCypher) [inline pkg/benchmarks/.eslintrc.js:8]: this is a literal explicit breaking of backwards compatability.

this is what you said you weren't going to do.

File: pkg/benchmarks/.npmrc
Comment (StoneCypher) [inline pkg/benchmarks/.npmrc:1]: this is a severe security error and must be undone

File: pkg/benchmarks/browser.js
Comment (StoneCypher) [inline pkg/benchmarks/browser.js:76]: benchmarks should not be dependant on network access to external files from a third party, especially not one with an outage history like github's.

this prevents people on corporate networks from participation.  historically this would have destroyed about 40% of peg's contributions from third parties.

unless the goal is to enforce that the public-facing examples in the repo run, or something?

but ... shouldn't that just be on the local copies?  that means that the local test set enforces a different version of the source than you're actually running it against

File: pkg/benchmarks/browser.js
Comment (StoneCypher) [inline pkg/benchmarks/browser.js:85]: as above, this should point at the local repo

the only thing hitting public URLs should be e2e tests, and these aren't the right URLs for those

File: pkg/benchmarks/browser.js
Comment (StoneCypher) [inline pkg/benchmarks/browser.js:99]: i don't understand what this does

File: pkg/benchmarks/package.json
Comment (StoneCypher) [inline pkg/benchmarks/package.json:24]: putting secondary packages into the main repo to create a monorepo is a classic security and maintainability disaster

there's literally never a good reason to do this.  this means that the subpackages aren't real packages anymore. 

want just one package?  nope, gotta download all of them.

want the simplicity of one piece?  nope, gotta understand all of them.

want standard tooling you're allowed to use in any corporate context?  nope, can't do this on your work pc at all.

literally no value is delivered here of any kind.  all this means is that the repo has a secondary alternate model that only a few people understand, and in order to contribute to peg at all, first you have to learn a completely different package management model, and sort out how the bizarro-land rube goldberg build system actually works

When futagoza did this three years ago, contributions hard-stopped immediately

please don't do this `:(`  

it is critical and important for public open source projects to work the regular way

this is a death knell

even high touch monorepos eventually get turned back normal because the dip in usage is too bad

this library is already almost dead and has already been dragged through three other unrelated monorepo concepts.  everyone's sick of it and a lot of old developers have already left in disgust, because the people taking over are wasting our time forcing us to learn their hipster tools

**Review by StoneCypher — CHANGES_REQUESTED** [review]
almost all of this work is specifically to create a monorepo under `pnpm` after you told me you weren't going to require any weird side novel package managers

this means nobody at `microsoft`, `google`, `apple`, `twitter`, or `netflix` can work on this anymore

please don't do this 😥  i told you before you started that this was a problem

i told futagoza before he started that this was a problem.  he did it anyway.  it was an instant full death for the library.

a bunch of third parties who had never been around before said 'oh this new shiny tool sounds great' and made him think it wasn't just going to drive everyone away

you have three years of people trying to force their need through, including me, so you'll get a little bit of traction, then it'll immediately go right back to being dead

all you're doing is modifying futagoza's damage.  this wasn't a monorepo before him.

please stop trying to make his damage stick.  monorepos are always bad for every project.

File: README.md
Comment (hildjj) [reply README.md:114]: I'm ok with that.  I'd like to document `es2015` or whatever we pick, but leave `es` as an alias, in case anyone started using it.

File: docs/development/test.html
Comment (hildjj) [reply docs/development/test.html:22]: I'll take a patch for that once this drops.  I'd rather spend time moving to whatever we pick that is not-mocha.

File: pkg/benchmarks/package.json
Comment (StoneCypher) [reply pkg/benchmarks/package.json:24]: i cannot point to a single monorepo across all of NPM that was popular and didn't immediately and permanently lose 50% of all its traffic

i tried to tell you "lerna is a nightmare" but you said "don't worry, i'm not using alternate package managers."

yes, you are.  

[i see you locked me out of my own issue just now](https://github.com/peggyjs/peggy/issues/76).

seems like you're being pretty explicit that even politely stated alternative ideas aren't welcome

Comment (StoneCypher) [reply pkg/benchmarks/package.json:24]: oh, nevermind, on reload there's a link to some new system called a discussion

i guess it doesn't put those there on hot-act like it does closures.  weird

File: package.json
Comment (hildjj) [reply package.json:2]: Example please?  `node-cbor` is a package that has this setup, and it seems to have been working ok.  Granted, it's not terribly popular compared to `peg.js`.

File: package.json
Comment (hildjj) [reply package.json:5]: yeah, the base package never gets published.

File: package.json
Comment (hildjj) [reply package.json:25]: There is a need to divide the base repo.  The CLI needs to take a node-only dependency to have any hope of being modern, and there's no reason for the web version to have a CLI or take a dependency.

File: package.json
Comment (hildjj) [reply package.json:35]: willfix

File: pkg/benchmarks/.eslintrc.js
Comment (hildjj) [reply pkg/benchmarks/.eslintrc.js:8]: All of the code that will be run on non-node12 will be run through babel first, so this will have no effect on backwards-compat.

File: pkg/benchmarks/.npmrc
Comment (hildjj) [reply pkg/benchmarks/.npmrc:1]: Since the core package has no dependencies, this just adds constant churn to track the dev tools.  We can talk about the cli.

File: pkg/benchmarks/browser.js
Comment (hildjj) [reply pkg/benchmarks/browser.js:76]: Nod, this is what the old benchmarks did.  Let's start a separate bug to fix this.

File: pkg/benchmarks/browser.js
Comment (hildjj) [reply pkg/benchmarks/browser.js:99]: this is adding links in the HTML so you can see what was being parsed, and is now likely to be incorrect.  Willfix after this lands and I can test it.

Comment (hildjj) [reply pkg/benchmarks/browser.js:99]: Again, this is just what was left over from before, it's not like this is code i'm proud of. :)

File: pkg/benchmarks/package.json
Comment (hildjj) [reply pkg/benchmarks/package.json:24]: I really don't understand your corporate context argument.  Can you say more about that please?  Assume in your answer that you'll be able to contribute just fine with npm once I remove the pnpm dependency.

Comment (StoneCypher) [reply pkg/benchmarks/package.json:24]: > I really don't understand your corporate context argument. Can you say more about that please? Assume in your answer that you'll be able to contribute just fine with npm once I remove the pnpm dependency.

Many can't install alternate package managers because corporate hasn't verified whether they do what they say on the box, and isn't willing to add them to the watchlist

Package managers have access to protected locations on disk that can run binaries.  They're higher stakes than typical JS libs.  Typical JS libs, you can just scan them to see if they have any on-install or on-build hooks then call it a day

**Comment by phpnode** [comment]
I don't feel as strongly about this as @StoneCypher does, and actually there are many examples of popular projects with monorepos, but I'm not sure about these splits in particular.

One of the things that I think is just incredibly important while Peggy is young is maintaining compatibility with pegjs so that we can get people to move over as simply as possible. At the moment it is effectively "find and replace pegjs with peggy and you're done". As soon as we start splitting CLI and library that process becomes (slightly) harder, and people might decide to stick where they are. It is more important to gain some momentum first imo.

If we do go with monorepo, then I'd like to keep the `peggy` package as similar to pegjs as possible. That would mean bundling the CLI and parser generator in the same library, but it would not prevent us from making a split under the hood. This could look like:

- `peggy` which composes:
  - `@peggyjs/peg` to turn a grammar into bytecode
  - `@peggyjs/cli` to run peggy from the command line
  - `@peggyjs/plugin-javascript` to generate javascript (or typescript / flow) from the bytecode

This would allow us to keep things familiar to people used to pegjs but also allow us to adopt a plugin architecture and gain the conveniences of a monorepo.

The other question I have is - are there remaining forks of pegjs which have useful work we'd like to include that would be harder to merge after a split like this? I know @mingun has done a lot of work in his fork for example. Maybe we should investigate that first?

**Comment by StoneCypher** [comment]
> actually there are many examples of popular projects with monorepos

Every single one of them saw a severe usage drop on entering monorepo, including the big names like React.  This is not coincidental.

Every time I've asked someone to provide an exception, they've been unable.

This happened to `peg` already.  😥

This hard-excludes corporate contributors, who are a large bulk of the past.  That on its own should be enough.

Unfortunately it seems like our dedication to compatibility is wavering before the first PR has landed.  

&nbsp;

> This would allow us to keep things familiar to people used to pegjs but also allow us to adopt a plugin architecture 

Respectfully, absolutely nothing about `pnpm` supports a plugin architecture in any way that wouldn't already exist under standard tooling.  By example, see several tools this repo is already using, such as webpack and eslint, which have plugins and no monorepos.

It's actually quite a bit more difficult, because now every plugin author has to start by learning `pnpm`, then by gaining rights to the main repo.

The main goal of most plugin architectures is to allow third parties to gladhandedly do things without the permission or interference of main.  Stuffing everything into a monorepo takes that away, and creates a bunch of bookkeeping in its place that doesn't serve the desired goal.

If you want a plugin architecture, monorepos work significantly against you.

&nbsp;

> are there remaining forks of pegjs which have useful work we'd like to include that would be harder to merge after a split like this?

Yes.  This throws away everything I did.

File: pkg/benchmarks/package.json
Comment (phpnode) [reply pkg/benchmarks/package.json:24]: in this scenario `pnpm` is installed via `npm`, it's not a global install like `yarn` would be. I think this should make it more palatable?

**Comment by StoneCypher** [comment]
The choice to use an alternative package manager is an extreme one.

You're cutting off a huge number of contributors.  

There does not seem to be a stated reason to actually do this.

The reasons given by third parties to justify the damage so far do not appear to actually hold water.  

Please don't shut me out this way.  There is no real-world upside here, and enormous real-world downside.

We've already made this mistake three other times.  It wasn't good any of those times.  We pretended it was about plugins those times too.  Zero plugins emerged.

I have a working plugin system that I'm ready to donate.  This makes it not work anymore.

**Comment by hildjj** [comment]
>    This could look like:
> 
>     * `peggy` which composes:
>       
>       * `@peggyjs/peg` to turn a grammar into bytecode
>       * `@peggyjs/cli` to run peggy from the command line
>       * `@peggyjs/plugin-javascript` to generate javascript (or typescript / flow) from the bytecode

That's a really cool idea.  However, I'm wavering on the whole idea.  Your argument about "search and replace peg.js for peggy" is compelling.  Maybe we can just live with a bad CLI forever and hope we never need runtime dependencies for anything.

**Comment by hildjj** [comment]
> The choice to use an alternative package manager is an extreme one.

I've already said I'll back that part out.

**Comment by StoneCypher** [comment]
As far as the CLI, what's the point of breaking it into another repo at all?

Nobody ships peg.  They ship what peg writes for them.  There's no binary size impact in a practical sense.

All we really need is a well written `.npmignore`

If you install peg, shouldn't it just have a working CLI?  That's how most `npm` projects work.  It's pretty rare to have to install a cli separately, and it's usually people trying to manage install size, which isn't relevant here

.

> `@peggyjs/peg` to turn a grammar into bytecode

b... bytecode?

wwwwwwwwwwwwwwwwhy?

are we going to write an entire compiler, instead of just working on what we already have?

just take the js output and compile it with an existing compiler

this is second system syndrome to an enormous degree.  

.

> * `@peggyjs/plugin-javascript` to generate javascript (or typescript / flow) from the bytecode

I don't even understand this thought process.

Is the idea that peg will generate some arbitrary bytecode, which will interact with a plugin system that knows that bytecode, and then be ... be transcompiled *up* into source languages we don't even originally support?

Who is going to write something like that?

Why would anyone want that?

Am I just misunderstanding? 

When we talk about peg plugins, it's generally expected to mean things that insert into peg, not things that insert into peg's newly created intermediate binary format

I mean GCC can't do this.  Is peg really going to do this?

Like why aren't we just doing source modules and parse transforms

It seems like this issue is being used as a place to suggest wild new conceptual features

File: package.json
Comment (StoneCypher) [reply package.json:2]: rollup 1.x

**Comment by phpnode** [comment]
@StoneCypher there must be some sort of misunderstanding yes, pegjs already generates bytecode, that bytecode is then transformed to JS by default. Plugins such as https://github.com/Nordth/php-pegjs and https://github.com/metadevpro/ts-pegjs then emit different output languages

File: package.json
Comment (StoneCypher) [reply package.json:5]: i really hope this comment predates where you said elsewhere that you were going to back `pnpm` back out.  

image

File: README.md
Comment (hildjj) [reply README.md:114]: Let's use #71 to track this since it's open and it's close enough to this discussion.  Please put your comment there?

File: docs/development/test.html
Comment (hildjj) [reply docs/development/test.html:22]: Start a new issue for this, please, if you want to track it.  Otherwise, resolve?

File: package.json
Comment (hildjj) [reply package.json:2]: I didn't follow the discussion they had.  Do you have a link, please?

File: package.json
Comment (StoneCypher) [reply package.json:25]: > There is a need to divide the base repo.

I do not agree.

&nbsp;

> The CLI needs to take a node-only dependency to have any hope of being modern

So what?  `package.json` is a node-only context.  dependencies from `package.json` don't affect non-cli things.

Just build the CLI in rollup.  It doesn't need access to dependencies after that.  This is a problem with webpack.

.

> there's no reason for the web version to have a CLI or take a dependency.

My core point is that there shouldn't be a "the web version"

The difference between the two is smaller than the packages being used to separate them

100% of the reason given for needing to split them is a result of having already split them

If you just don't split them in the first place, the reason goes away

This would be much simpler in the original single-repo non-monorepo structure

File: package.json
Comment (hildjj) [reply package.json:5]: Not necessarily, but I could see a compromise where what is currently in `pkg/peggy` moves down into the root directory.  I bet that would fix enough of the things that are bothering you that we could reset the discussion, so I'll do that work after lunch (MDT).

File: package.json
Comment (StoneCypher) [reply package.json:25]: This didn't exist at all before futagoza

It worked a lot better back then

This was added because futagoza tried to move the hosting to some knockoff of heroku that made deploys using the git repo as structure, then later off of it when they went out of business

File: pkg/benchmarks/.eslintrc.js
Comment (StoneCypher) [reply pkg/benchmarks/.eslintrc.js:5]: ok

File: pkg/benchmarks/.eslintrc.js
Comment (StoneCypher) [reply pkg/benchmarks/.eslintrc.js:8]: Oh god babel?

1. A lot of babel's shims are just incorrect
2. You can expect the library to slow down ~300%
3. Babel introduces a lot of undefined behavior instead of throwing when it hits undefined behavior.  These kinds of bugs are essentially impossible to locate
4. If it's being transcompiled, can we use `typescript` instead?  
    1. We can start with all checks off, so it won't slow you down
    2. It will immediately add significant value in noticing things like missing arguments
    3. Its shims are far faster and much more commonly correct
    4. You get much better language support
    5. We have the option of stepping into a typesystem progressively with time
    6. We need it anyway for dropping .d.ts
    7. Everybody can help (babel is an increasingly rare skill outside the react ecosystem, and the react ecosystem is mostly moved to typescript at this point)

&nbsp;

> this will have no effect on backwards-compat.

Babel's shims clash violently with modernizr and a lot of peg people use modernizr

File: pkg/benchmarks/browser.js
Comment (StoneCypher) [reply pkg/benchmarks/browser.js:76]: Oh okay.  I hadn't looked at them before.

[Made](https://github.com/peggyjs/peggy/issues/78)

File: package.json
Comment (hildjj) [reply package.json:25]: > Just build the CLI in rollup. It doesn't need access to dependencies after that.

Now *that* is a good idea.

File: pkg/benchmarks/browser.js
Comment (StoneCypher) [reply pkg/benchmarks/browser.js:99]: Got it.  [Same issue then](https://github.com/peggyjs/peggy/issues/78)

File: pkg/benchmarks/package.json
Comment (StoneCypher) [reply pkg/benchmarks/package.json:24]: > in this scenario pnpm is installed via npm, it's not a global install like yarn would be. I think this should make it more palatable?

Until the security team looks it over, which will probably be never, unfortunately no.

There's a cost to verifying whether a tool is still safe, on each release. It's hard to get an employer to commit to adding a name to that list.

Comment (StoneCypher) [reply pkg/benchmarks/package.json:24]: > if we move forward I'll remove pnpm as a dependency and from being called in the scripts.

awesome, thank you

sorry to be weird about it

File: pkg/benchmarks/run
Comment (StoneCypher) [inline pkg/benchmarks/run:36]: i realize this is old code that didn't come from you and i'm just seeing it because of the `let` -> `const` switch, but, also this is weird code and i don't like it.  there's all sorts of bizarre failure cases, like if length is shorter than the source string, that are just not failing by chance

```typescript
function center(text, length) {

  const emptyLength = Math.max(length - text.length, 0),
        leftPad     = Math.floor(emptyLength / 2),
        rightPad    = emptyLength - leftPad;

  return (' '.repeat(leftPad)) + (text) + (' '.repeat(rightPad));

}
```

**Review by StoneCypher — COMMENTED** [review]
> pegjs already generates bytecode

oh.  jesus this whole time i thought it was an AST

File: README.md
Comment (StoneCypher) [reply README.md:114]: done

File: docs/development/test.html
Comment (StoneCypher) [reply docs/development/test.html:22]: these are comments, not reviews, so i think that they can't be resolved

i could be wrong

[issue is here](https://github.com/peggyjs/peggy/issues/78)

Comment (hildjj) [reply docs/development/test.html:22]: Ah, I might have to be the one to resolve it.

File: package.json
Comment (StoneCypher) [reply package.json:2]: i'm not aware of a discussion around this.  it's just something i went through years ago

File: package.json
Comment (StoneCypher) [reply package.json:5]: > Also, can I ask you respectfully to remove the meme above? I don't think it's going to help this conversation resolve more quickly.

i guess

memes make it safe to show distress in a way that other people don't find to be aggressive

File: package.json
Comment (StoneCypher) [reply package.json:25]: I'll do the labor if you want, have it running a bunch of other places too already

File: pkg/peggy/lib/version.js
Comment (StoneCypher) [inline pkg/peggy/lib/version.js:3]: it's probably best to generate this from the root package

i have standard stuff for this that i can donate if youw ant it

**Comment by hildjj** [comment]
OK, I'm sitting down to refactor this some.  I fully expect that even after this refactor, we'll go with the approach of building the CLI with rollup instead of this, but I'm going to do the work anyway as a show of good faith.

File: pkg/peggy/lib/version.js
Comment (hildjj) [reply pkg/peggy/lib/version.js:3]: yes please.

Comment (hildjj) [reply pkg/peggy/lib/version.js:3]: once we settle this down so you have a clean place to PR from.

**Comment by StoneCypher** [comment]
happy to pitch in as you see fit

**Comment by hildjj** [comment]
Here's roughly what I meant.  This will kind of suck to work on without pnpm or yarn, however, because npm's `workspaces` is exactly backwards of what it should be, imo.

so: let's punt on this, and start over.  I think the first thing is to switch from browserify to rollup, instead of webpack.

**Comment by StoneCypher** [comment]
oh, mylanta, @phpnode is right; it came up in a different bug

```
        28,                          // SILENT_FAILS_ON
        18, 1, 2, 2, 22, 1, 23, 2,   // <expression>
        29,                          // SILENT_FAILS_OFF
        14, 2, 0,                    // IF_ERROR
        23, 0                        //   * FAIL
```

this is the worst time to not be allowed to use memes

**Comment by StoneCypher** [comment]
> so: let's punt on this, and start over. I think the first thing is to switch from browserify to rollup, instead of webpack.

This works for me

I would like to appeal for a series of small pieces, rather than one large piece

Do you want me to do this step and/or help with this step?

**Comment by hildjj** [comment]
Yes, I'd like help.  I'll start a new bug to move to rollup, and assign it to you, @StoneCypher.  That's going to block a bunch of other work, but I'll start working off some of the other minor issues in the meantime.

**Comment by hildjj** [comment]
See #83

**Pull request closed without merging**
